src/foam/core/Promised.js

```diff
@@ -129,38 +109,35 @@ foam.CLASS({
       var myAxioms = [
         foam.core.Proxy.create({
           name: stateName,
-          flags: ['js', 'swift'],
           of: this.of,
-          delegates: methodNames,
-          forwards: [],
+          forwards: methodNames,
           factory: function() {
-            return this[pendingState].create();
+            return this[pendingState].create({obj: this});
           },
-          swiftFactory: `return ${pendingState}_create(["obj": self])`,
+          swiftFactory: `
+            return ${pendingState}_create(["obj": self])
+          `,
+          javaFactory: `
+            return new ${pendingState}.Builder(getX()).setObj(this).build();
+          `,
           transient: true
         }),
-        foam.core.Property.create({
-          name: delegateName,
-          flags: ['js', 'swift'],
-          postSet: function() {
-            this[stateName] = this[fulfilledState].create();
-          },
-          swiftGetter: `return try! ${myName}.get() as! ${foam.lookup(this.of).model_.swiftName}`,
-        }),
         foam.core.ProxySub.create({
           topics: this.topics,
-          prop:   delegateName
+          prop:   stateName
         })
       ];
 
       var pendingMethods = [];
 
+      // Use all methods from here on in to satisfy the interface.
+      // Methods that aren't part of this.methods should never get called
+      // but they need to be added to make compilers happy.
+      methods = delegate.getOwnAxiomsByClass(foam.core.Method);
       for ( var i = 0 ; i < methods.length ; i++ ) {
         pendingMethods.push(foam.core.PromisedMethod.create({
           name: methods[i].name,
-          flags: ['js', 'swift'],
           property: myName,
-          type:  methods[i].type,
```

Comment:
Why drop the type?

Reply:
Since the model that we're applying to implements the interface of the `of`, the types and args are pulled in automatically.